File: Lista de Exercicio Paradigmas/Lista Python/Ex9.py
from abc import ABC, abstractmethod

class Imprimivel(ABC):
    @abstractmethod
    def imprimir(self):
        pass

class Relatorio(Imprimivel):
    def __init__(self, conteudo):
        self.conteudo = conteudo

    def imprimir(self):
        print(f"Imprimindo Relatório:\n{self.conteudo}")

class Contrato(Imprimivel):
    def __init__(self, partes):
        self.partes = partes

    def imprimir(self):
        print("Imprimindo Contrato:")
        for parte in self.partes:
            print(f"- {parte}")

# Exemplo de uso
relatorio = Relatorio("Este é um relatório de vendas.")
contrato = Contrato(["Parte A: João", "Parte B: Maria"])

relatorio.imprimir()
contrato.imprimir()

File: Lista de Exercicio Paradigmas/Lista Java/Ex9.java
import java.util.List;

abstract class Imprimivel {
    abstract void imprimir();
}

class Relatorio extends Imprimivel {
    private String conteudo;

    public Relatorio(String conteudo) {
        this.conteudo = conteudo;
    }

    @Override
    void imprimir() {
        System.out.println("Imprimindo Relatório:\n" + this.conteudo);
    }
}

class Contrato extends Imprimivel {
    private List<String> partes;

    public Contrato(List<String> partes) {
        this.partes = partes;
    }

    @Override
    void imprimir() {
        System.out.println("Imprimindo Contrato:");
        for (String parte : partes) {
            System.out.println("- " + parte);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Relatorio relatorio = new Relatorio("Este é um relatório de vendas.");
        Contrato contrato = new Contrato(List.of("Parte A: João", "Parte B: Maria"));

        relatorio.imprimir();
        contrato.imprimir();
    }
}
